### PR TITLE
NavigationHeader

### PR DESCRIPTION
**Pull request by dephillipsmichael**
Added method for sub-classes to hook into last vertical view

### CODE REVIEW

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 1]
```diff
@@ -442,6 +442,10 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
         setupVerticalConstraints(textLabel)
         setupVerticalConstraints(learnMoreButton)
         
+        if let finalVerticalView = setupFinalVerticalViews() {
+            lastView = finalVerticalView
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: It seems like you're going to want to be able to append some NSLayoutConstraints to  _interactiveConstraints for the finalVerticalView, or this could end up with layout issues--but subclasses won't have access to that private var so you'll need to add a method to this class that lets subclasses do that.

Also, since this changes the value of lastView, which is used in setupVerticalConstraints(...) above, shouldn't this happen *before* all those calls?

Also also shouldn't there be a call to setupVerticalConstraints(finalVerticalView)?

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 2]
```diff
@@ -473,6 +477,12 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
         }
     }
     
+    /// This function is available to sub-classes to append additional views as the last view in the navigation header
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: needs a period at the end.

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 3]
```diff
@@ -473,6 +477,12 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
...
+    /// @return the final view that will be bound to the bottom of the navigation header superview, nil if none were added
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: @return --> - returns:

Also initial 'The' should be capitalized, needs an 'or' before 'nil', and needs a period at the end.

**Review by Erin-Mounts — CHANGES_REQUESTED** [review]

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift (hunk 1 above)
Comment (dephillipsmichael) [reply ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: For more context, this is my subclass in mPower app

open class SBATrackedMedicationNavigationHeaderView: RSDTableStepHeaderView {
    
    private var _underlinedButtonContraints: [NSLayoutConstraint] = []
    
    /// The label for displaying step detail text.
    @IBOutlet open var underlinedButton: RSDUnderlinedButton?
    
    public override init() {
        super.init(frame: CGRect.zero)
        commonInitNavigationHeaderView()
    }
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        commonInitNavigationHeaderView()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInitNavigationHeaderView()
    }
    
    private func commonInitNavigationHeaderView() {
        addUnderlinedButtonIfNeeded()
    }
    
    override open func setupFinalVerticalViews() -> UIView? {
        // Remove existing constraints for the underlined button
        NSLayoutConstraint.deactivate(_underlinedButtonContraints)
        _underlinedButtonContraints.removeAll()
        
        // Add the underlined button under the title label
        if let underlinedButtonUnwrapped = self.underlinedButton,
            let titleLabelUnwrapped = super.titleLabel {
            _underlinedButtonContraints.append(contentsOf:
                underlinedButtonUnwrapped.rsd_alignBelow(view: titleLabelUnwrapped, padding: constants.verticalSpacing))
        }
        return underlinedButton
    }
    
    /// Convenience method for adding a label.
    open func addUnderlinedButton(font: UIFont, color: UIColor) -> RSDUnderlinedButton {
        let button = RSDUnderlinedButton(type: .system)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.titleLabel?.font = font
        button.setTitleColor(color, for: .normal)
        button.titleLabel?.textAlignment = .center
        button.titleLabel?.preferredMaxLayoutWidth = constants.labelMaxLayoutWidth
        self.addSubview(button)
        
        button.rsd_alignToSuperview([.leading, .trailing], padding: constants.sideMargin)
        button.rsd_makeHeight(.greaterThanOrEqual, 60.0)
        
        return button
    }
    
    func addUnderlinedButtonIfNeeded() {
        guard underlinedButton == nil else { return }
        underlinedButton = addUnderlinedButton(font: UIFont.rsd_headerTextLabel, color: UIColor.rsd_headerTextLabel)
    }
}

Comment (Erin-Mounts) [reply ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: Discussed in Slack:
themdp [5:31 PM]
[...]
the problem is, the header calculates it height through autolayout and it does that by setting the last view's bottom constraint to superview
so i needed a way to inject a view as the last view
in mpower, i did this...

themdp [5:33 PM]
added this Java snippet: Untitled 
open class SBATrackedMedicationNavigationHeaderView: RSDTableStepHeaderView {
  
  private var _underlinedButtonContraints: [NSLayoutConstraint] = []
  
  /// The label for displaying step detail text.
  @IBOutlet open var underlinedButton: RSDUnderlinedButton?
  
  public override init() {
    super.init(frame: CGRect.zero)
    commonInitNavigationHeaderView()
  }
  
  public override init(frame: CGRect) {
    super.init(frame: frame)
    commonInitNavigationHeaderView()
  }
  
  required public init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
    commonInitNavigationHeaderView()
  }
  
  private func commonInitNavigationHeaderView() {
    addUnderlinedButtonIfNeeded()
  }
  
  override open func setupFinalVerticalViews() -> UIView? {
    // Remove existing constraints for the underlined button
    NSLayoutConstraint.deactivate(_underlinedButtonContraints)
    _underlinedButtonContraints.removeAll()
    
    // Add the underlined button under the title label
    if let underlinedButtonUnwrapped = self.underlinedButton,
      let titleLabelUnwrapped = super.titleLabel {
      _underlinedButtonContraints.append(contentsOf:
        underlinedButtonUnwrapped.rsd_alignBelow(view: titleLabelUnwrapped, padding: constants.verticalSpacing))
    }
    return underlinedButton
  }
  
  /// Convenience method for adding a label.
  open func addUnderlinedButton(font: UIFont, color: UIColor) -> RSDUnderlinedButton {
    let button = RSDUnderlinedButton(type: .system)
    button.translatesAutoresizingMaskIntoConstraints = false
    button.titleLabel?.font = font
    button.setTitleColor(color, for: .normal)
    button.titleLabel?.textAlignment = .center
    button.titleLabel?.preferredMaxLayoutWidth = constants.labelMaxLayoutWidth
    self.addSubview(button)
    
    button.rsd_alignToSuperview([.leading, .trailing], padding: constants.sideMargin)
    button.rsd_makeHeight(.greaterThanOrEqual, 60.0)
    
    return button
  }
  
  func addUnderlinedButtonIfNeeded() {
    guard underlinedButton == nil else { return }
    underlinedButton = addUnderlinedButton(font: UIFont.rsd_headerTextLabel, color: UIColor.rsd_headerTextLabel)
  }
}
Collapse 

themdp [5:34 PM]
i basically keep my own _interactiveConstraints as a var and keep track of those
sorry i should be posting this in the PR

erinmounts-sage [5:38 PM]
got it, that makes sense since it gets called whenever the updateInteractiveConstraints gets called

themdp [5:40 PM]
Based on the subclass context, is there still a more appropriate way I could go about this?

erinmounts-sage [5:41 PM]
I mean, if there’s a way to do it by overriding the public updateConstraints() func and calling super at the appropriate time, that would be ideal

themdp [5:43 PM]
That was my initial thought.  But, that would require me to somehow remove only the bottomConstraint of the lastView, which i would have no idea how to do.
does that make sense?

erinmounts-sage [5:47 PM]
I am trying to come up with a “cleaner” alternative and I think I am just recapitulating your thought process and coming to the same conclusions

themdp [5:47 PM]
lol
ive been there on a pr review
I think you were on to something though with your comment about setupVertical constraint
I think we could just put this in its own method
func setupVerticalConstraints(_ nextView: UIView?) {
           if let vw = nextView, shouldLayout(vw) {
               applyVerticalConstraint(to: vw, lastView: lastView)
               firstView = firstView == nil ? vw : firstView
               lastView = vw
           } else {
               nextView?.isHidden = true
           }
       }
       
       // image view
       setupVerticalConstraints(imageView)
       setupVerticalConstraints(titleLabel)
       setupVerticalConstraints(textLabel)
       setupVerticalConstraints(learnMoreButton)
and overriding that wouldnt accomplish my goal i think in a cleaner way
let me see if I can do it

erinmounts-sage [5:51 PM]
hm, yes that does feel a bit cleaner, and should be equivalent--please document that method as a suitable place to override to add more views to the vertical stack (seems like you should call super *after* injecting your views, but I’m not 100% positive about that)

themdp [5:52 PM]
The order of constraint additions doesnt matter, just their priority value
So I think you could do either way

erinmounts-sage [5:52 PM]
well since it depends on the value of lastView
and you’re changing lastView

themdp [5:53 PM]
Good thought, I'll look into that more

**Comment by dephillipsmichael** [comment]
New subclass implementation:

open class SBATrackedMedicationNavigationHeaderView: RSDTableStepHeaderView {
    
    private var _underlinedButtonContraints: [NSLayoutConstraint] = []
    
    var delegate: SBATrackedMedicationNavigationHeaderViewDelegate?
    var underlinedButtonText: String?
    /// The label for displaying step detail text.
    @IBOutlet open var underlinedButton: RSDUnderlinedButton?
    
    override open func updateVerticalConstraints(currentLastView: UIView?) -> (firstView: UIView?, lastView: UIView?) {
        addUnderlinedButtonIfNeeded()
        var results = super.updateVerticalConstraints(currentLastView: currentLastView)
        
        // Remove existing constraints for the underlined button
        NSLayoutConstraint.deactivate(_underlinedButtonContraints)
        _underlinedButtonContraints.removeAll()
        
        // Add the underlined button under the title label
        if let underlinedButtonUnwrapped = self.underlinedButton,
            let titleLabelUnwrapped = super.titleLabel {
            _underlinedButtonContraints.append(contentsOf:
                underlinedButtonUnwrapped.rsd_alignBelow(view: titleLabelUnwrapped, padding: constants.verticalSpacing))
        }
        
        results.lastView = underlinedButton
        return results
    }
    
    func addUnderlinedButtonIfNeeded() {
        guard underlinedButton == nil else { return }
        underlinedButton = addUnderlinedButton(font: UIFont.rsd_headerTextLabel, color: UIColor.rsd_headerTextLabel)
    }
    
    /// Convenience method for adding an underlined button.
    open func addUnderlinedButton(font: UIFont, color: UIColor) -> RSDUnderlinedButton {
        let button = RSDUnderlinedButton(type: .system)
        button.translatesAutoresizingMaskIntoConstraints = false
        button.titleLabel?.font = font
        button.setTitleColor(color, for: .normal)
        button.titleLabel?.textAlignment = .center
        button.titleLabel?.preferredMaxLayoutWidth = constants.labelMaxLayoutWidth
        button.setTitle(underlinedButtonText, for: .normal)
        button.addTarget(self, action: #selector(self.underlineButtonTapped), for: .touchUpInside)
        self.addSubview(button)
        
        button.rsd_alignToSuperview([.leading, .trailing], padding: constants.sideMargin)
        button.rsd_makeHeight(.greaterThanOrEqual, 60.0)
        
        return button
    }
    
    @objc func underlineButtonTapped() {
        self.delegate?.underlinedButtonTapped()
    }
}

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 4]
```diff
@@ -473,6 +461,40 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
         }
     }
     
+    /// Set up the vertical constraints for the views that should be stacked vertically, like the imageView, titleLabel, textLabel, and learnMoreButton. You will need to remove previous constraints and re-add the ones associated with your vertically stacked view.
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: I'd put the second sentence on its own line. Also I'd reword it:
```
/// Your subclass can override this function to add more vertically stacked views, either before or after all the
/// existing ones, depending on whether you add them before or after calling the `super` function. If you do, 
/// you'll need to manage removing and re-adding the associated constraints in your override as well.
```

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 5]
```diff
@@ -473,6 +461,40 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
...
+    /// @param currentLastView the lastView that was added to the navigation header, if no views are added in this function, this should be returned.
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]:
```
/// - parameter currentLastView: The last (vertical) view currently in the navigation header. If no views are
///                              added in a subclass override of this function, this should be returned in the
///                              lastView parameter.
```

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 6]
```diff
@@ -473,6 +461,40 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
...
+    /// @return (firstView, lastView) firstView that had vertical constraints applied, nil if none were applied
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]:
```
/// - returns:  A tuple with the firstView and lastView that had vertical constraints applied. If none were
///             applied, firstView will be nil, and lastView will be the same as currentLastView.

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 7]
```diff
@@ -473,6 +461,40 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
...
+    ///                               lastView that had vertical constraints applied, nil if none were applied
+    open func updateVerticalConstraints(currentLastView: UIView?) -> (firstView: UIView?, lastView: UIView?) {
+        
+        var firstView: UIView?
+        var lastView: UIView?
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: Should default to currentLastView:
`    var lastView: UIView? = currentLastView`

File: ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift [hunk 8]
```diff
@@ -473,6 +461,40 @@ open class RSDTableStepHeaderView: RSDStepHeaderView {
...
+        // image view
+        setupVerticalConstraints(imageView)
+        setupVerticalConstraints(titleLabel)
+        setupVerticalConstraints(textLabel)
+        setupVerticalConstraints(learnMoreButton)
+        
+        return (firstView, lastView)
+    }
+    
+    /// This function is available to sub-classes to append additional views as the last view in the navigation header
```
Comment (Erin-Mounts) [inline ResearchUI/ResearchUI/iOS/Views/RSDStepNavigationView.swift]: this func is no longer called and can be removed

**Review by Erin-Mounts — CHANGES_REQUESTED** [review]

**Review by Erin-Mounts — APPROVED** [review]
looks good, thanks!
